cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -175,42 +182,231 @@ size_type inplace_bitmask_binop(
+                                              stream.value()>>>(
+    bitmask, num_ranges, first_bit_indices, last_bit_indices, d_null_counts.begin());
+
+  if (count_bits == count_bits_policy::UNSET_BITS) {
+    // Subtract the number of set bits from the length of the segment
+    thrust::for_each(rmm::exec_policy(stream),
+                     thrust::make_counting_iterator(0),
+                     thrust::make_counting_iterator(static_cast<size_type>(d_null_counts.size())),
+                     [first_bit_indices,
+                      last_bit_indices,
+                      d_null_counts = d_null_counts.data()] __device__(size_type i) {
+                       auto const begin = *(first_bit_indices + i);
+                       auto const end   = *(last_bit_indices + i);
+                       d_null_counts[i] = (end - begin) - d_null_counts[i];
+                     });
+  }
+
+  CHECK_CUDA(stream.value());
+  return d_null_counts;
+}
+
+struct index_alternator {
+  CUDA_DEVICE_CALLABLE size_type operator()(const size_type& i) const
+  {
+    return *(d_indices + 2 * i + (is_end ? 1 : 0));
+  }
+
+  bool const is_end = false;
+  const size_type* d_indices;
+};
+
+/**
+ * @brief Given a bitmask, counts the number of set (1) or unset (0) bits in every range
  * `[indices_begin[2*i], indices_begin[(2*i)+1])` (where 0 <= i < std::distance(indices_begin,
  * indices_end) / 2).
  *
- * Returns an empty vector if `bitmask == nullptr`.
+ * If `bitmask == nullptr`, this function returns a vector containing the
+ * segment lengths, or a vector of zeros if counting unset bits.
  *
  * @throws cudf::logic_error if `std::distance(indices_begin, indices_end) % 2 != 0`
  * @throws cudf::logic_error if `indices_begin[2*i] < 0 or indices_begin[2*i] >
  * indices_begin[(2*i)+1]`
  *
  * @param bitmask Bitmask residing in device memory whose bits will be counted
  * @param indices_begin An iterator representing the beginning of the range of indices specifying
- * ranges to count the number of set bits within
+ * ranges to count the number of set/unset bits within
  * @param indices_end An iterator representing the end of the range of indices specifying ranges to
- * count the number of set bits within
+ * count the number of set/unset bits within
+ * @param count_bits If SET_BITS, count set (1) bits. If UNSET_BITS, count unset (0) bits
  * @param stream CUDA stream used for device memory operations and kernel launches
  *
  * @return A vector storing the number of non-zero bits in the specified ranges
  */
 template <typename IndexIterator>
-std::vector<size_type> segmented_count_set_bits(bitmask_type const* bitmask,
-                                                IndexIterator indices_begin,
-                                                IndexIterator indices_end,
-                                                rmm::cuda_stream_view stream)
+std::vector<size_type> segmented_count_bits(bitmask_type const* bitmask,
```

Comment:
It's fine to use a flag for the internal implementation, but for the user-facing API I would prefer separately named functions that just call the implementation with the different `count_bits_policy`.

Reply:
I think this might already be designed as you're describing? The public API has functions `segmented_count_set_bits` and `segmented_count_unset_bits`. Only `detail` APIs use the `count_bits_policy`. https://github.com/rapidsai/cudf/blob/2f76450afcbf57534c73666c9b44616ab838c665/cpp/src/bitmask/null_mask.cu#L471-L487

Reply:
I guess by "user facing" I meant users of this function internal to libcudf, i.e., using the iterator based `detail::` algorithm.